lib/value/src/value/iter.rs

```diff
@@ -0,0 +1,466 @@
+    ///   elements within the array.
+    /// - Any mutations done to the array before recursing into it are
+    ///   preserved, meaning once recursion starts, the mutations done on the
+    ///   object itself are preserved.
+    pub fn into_iter<'a>(self, recursive: bool) -> ValueIter<'a> {
+        let data = match self {
+            Value::Object(object) => IterData::Object(object.into_iter().collect()),
+            Value::Array(array) => IterData::Array(array),
+            value => IterData::Value(value),
+        };
+
+        ValueIter::new(data, recursive)
+    }
+}
+
+/// An [`Iterator`] over a [`Value`].
+pub struct ValueIter<'a> {
+    data: IterData,
+    recursive: bool,
+    index: usize,
+    recursive_iter: Option<Box<Self>>,
+    must_prepare_recursion: bool,
+    phantom: PhantomData<&'a mut ()>,
+}
+
+/// The [`Iterator::Item`] returned by the [`ValueIter`] iterator.
+#[derive(Debug, PartialEq, Eq)]
+#[allow(clippy::enum_variant_names)]
+pub enum IterItem<'a> {
+    /// A single primitive value.
+    Value(&'a mut Value),
+
+    /// A key/value combination.
+    KeyValue(&'a mut String, &'a mut Value),
+
+    /// An index/value combination.
+    IndexValue(usize, &'a mut Value),
+}
+
+/// The internal data representation used by the iterator.
+///
+/// Objects are stored as a `Vec<(String, Value)>`, to allow mutating the object
+/// keys during iteration.
+enum IterData {
+    Value(Value),
+    Object(Vec<(String, Value)>),
+    Array(Vec<Value>),
+}
+
+impl<'a> ValueIter<'a> {
+    /// Create a new iterator over the relevant [`IterData`].
+    fn new(data: IterData, recursive: bool) -> Self {
+        Self {
+            data,
+            recursive,
+            index: 0,
+            recursive_iter: None,
+            must_prepare_recursion: false,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a> Iterator for ValueIter<'a> {
+    type Item = IterItem<'a>;
+
+    #[allow(clippy::deref_addrof)]
+    fn next(&mut self) -> Option<Self::Item> {
+        // If this returns true, it means on the last iteration cycle, we've
+        // returned a collection value-type, and the caller asked to recurse
+        // into that collection.
+        //
+        // We're going to prepare recursion by embedding a new recursive
+        // iterator into `ValueIter`. This preparation is delayed to this new
+        // cycle, to allow the caller to first mutate the collection we're going
+        // to recurse into.
+        if self.must_prepare_recursion {
+            self.must_prepare_recursion = false;
+
+            // Get the relevant collection `Value` type we want to iterate over.
+            let value = match &mut self.data {
+                IterData::Object(object) => &mut object.get_mut(self.index - 1)?.1,
+                IterData::Array(array) => array.get_mut(self.index - 1)?,
+                IterData::Value(_) => unreachable!("cannot recurse into non-container type"),
+            };
```

Comment:
Could the `match value` shared code be moved into a function such that this preparation could be moved into each arm of the `match &mut self.data` and avoid this extra `match` block here?